tools: Enable Python bridge on Fedora 38

martinpitt opened the pull request:
After this, we can remove the fedora-38/pybridge scenario.

 - [x] PR #19025
 - [x] PR #19029
 - [x] PR #19034
 - [x] PR #19038
 - [x] more properly fix package loading (@allisonkarlitskaya is working on it) #19065  ? Alternatively I added a commit to avoid the  [OOM failure](https://cockpit-logs.us-east-1.linodeobjects.com/pull-19009-20230713-053600-3a36307a-fedora-38-devel/log.html)

martinpitt commented:
Oh, this is interesting -- it also enables the pybridge on Fedora CoreOS, which apparently needs some work.

martinpitt commented:
Could you please do an initial review of this? I think I want to clean up the "memory eater" commit more, but I'd appreciate @allisonkarlitskaya 's gut feeling/judgement about that. Also, @mvollmer, I suppose there was some particular reason for the `init` name of the superuser? Can you please have an initial look?

Thanks!

allisonkarlitskaya commented on pyproject.toml:
```diff
@@ -56,6 +56,7 @@ ignore = [
     "A003",  # Class attribute is shadowing a python builtin
     "B011",  # Do not `assert False` (`python -O` removes these calls), raise `AssertionError()`
     "E731",  # Do not assign a `lambda` expression, use a `def`
+    "FBT002",  # Boolean default value in function definition
```
Why?

allisonkarlitskaya commented on src/cockpit/packages.py:
```diff
@@ -317,11 +319,13 @@ def get_content_security_policy(self, origin):
             # Otherwise, just look up the file based on its path.  May raise KeyError.
-            data, (content_type, encoding) = self.files[path]
+            path, (content_type, encoding) = self.files[path]
+            data = path.read_bytes()
```
Please make sure you also handle `OSError` from the same place that `KeyError` is handled in `Packages.serve_package_file()`.  It might even make sense to return different messages depending on what is wrong.

martinpitt replied:
Done.

allisonkarlitskaya commented on src/cockpit/superuser.py:
```diff
@@ -168,19 +168,19 @@ async def go(self, name: str, responder: ferny.InteractionResponder) -> None:
             raise bus.BusError('cockpit.Superuser.Error', f'Unknown superuser bridge type "{name}"') from exc
 
-        self.current = 'init'
+        self.current = name
```
The reason we don't do this from the start is because we don't know that it will succeed, and we want the UI to remain in an indeterminate state until we have confirmation one way or the other.

Also: the justification in the commit message is confusing.  Package loading is synchronous and the bridges should be fully configured before `Bridge.__init__` completes and the bridge starts processing messages.  Are we somehow doing an accidental reload on login?

martinpitt commented:
I took a look at the [fedora-coreos/expensive](https://cockpit-logs.us-east-1.linodeobjects.com/pull-19009-20230625-091951-4674125e-fedora-coreos-expensive/log.html) and [fedora-coreos/other](https://cockpit-logs.us-east-1.linodeobjects.com/pull-19009-20230625-094858-4674125e-fedora-coreos-other/log.html) failures. check-system-info and check-pages failures are fixed now with PR #19025.

All check-shell-* still happen and need investigation.

martinpitt commented:
Fixed the conflict and dropped the [memory saver patch](https://github.com/cockpit-project/cockpit/commit/8c3e23a602f40557f0f01ca7975c68815f5bc787) for now. This will most probably fail the /devel scenario again, but let's make sure everything else still works.

martinpitt commented:
The test actually [looks a lot better now](https://cockpit-logs.us-east-1.linodeobjects.com/pull-19009-20230713-053600-3a36307a-fedora-38-devel/log.html), as commit 1e191f94d66bc4b985 already removed half of the RAM usage.

So my proposal is to work around that by giving the provisioned machine a bit more memory than 512 MiB in the devel scenario. This shouldn't be a big practical problem as real users are going to use the small production version of the pages. That would allow us to get some more field testing. @allisonkarlitskaya what is your gut feeling, should we still block this on your packages.py rework, or are you okay with unleashing it with the next release in two weeks?

allisonkarlitskaya commented:
Give me another day on the packaging rework (and maybe start initially reviewing).  It's not far now.

allisonkarlitskaya reviewed (dismissed):
:rocket:

allisonkarlitskaya reviewed (approved):
> This is a known flake, retrying.

aww too bad :)

+1, then.
